pkg/encryption/store.go

```diff
@@ -0,0 +1,206 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package encryption
+
+import (
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/paths"
+	"storj.io/storj/pkg/storj"
+)
+
+// The Store allows one to find the matching most encrypted key and path for
+// some unencrypted path. It also reports a mapping of encrypted to unencrypted paths
+// at the searched for unencrypted path.
+//
+// For example, if the Store contains the mappings
+//
+//    b1, u1/u2/u3    => <e1/e2/e3, k3>
+//    b1, u1/u2/u3/u4 => <e1/e2/e3/e4, k4>
+//    b1, u1/u5       => <e1/e5, k5>
+//    b1, u6          => <e6, k6>
+//    b1, u6/u7/u8    => <e6/e7/e8, k8>
+//    b2, u1          => <e1', k1'>
+//
+// Then the following lookups have outputs
+//
+//    b1, u1          => <{e2:u2, e5:u5}, u1, nil>
+//    b1, u1/u2/u3    => <{e4:u4}, u1/u2/u3, <u1/u2/u3, e1/e2/e3, k3>>
+//    b1, u1/u2/u3/u6 => <{}, u1/u2/u3/, <u1/u2/u3, e1/e2/e3, k3>>
+//    b1, u1/u2/u3/u4 => <{}, u1/u2/u3/u4, <u1/u2/u3/u4, e1/e2/e3/e4, k4>>
+//    b1, u6/u7       => <{e8:u8}, u6/, <u6, e6, k6>>
+//    b2, u1          => <{}, u1, <u1, e1', k1'>>
+type Store struct {
+	roots      map[string]*node
+	defaultKey *storj.Key
+}
+
+// node is a node in the Store graph. It may contain an encryption key and encrypted path,
+// a list of children nodes, and data to ensure a bijection between encrypted and unencrypted
+// path entries.
+type node struct {
+	unenc    map[string]*node  // unenc => node
+	unencMap map[string]string // unenc => enc
+	enc      map[string]*node  // enc => node
+	encMap   map[string]string // enc => unenc
+	base     *Base
+}
+
+// Base represents a key with which to derive further keys at some encrypted/unencrypted path.
+type Base struct {
+	Unencrypted paths.Unencrypted
+	Encrypted   paths.Encrypted
+	Key         storj.Key
+}
+
+// clone returns a copy of the Base. The implementation can be simple because the
+// types of its fields do not contain any references.
+func (b *Base) clone() *Base {
+	if b == nil {
+		return nil
+	}
+	bc := *b
+	return &bc
+}
+
+// NewStore constructs a Store.
+func NewStore() *Store {
+	return &Store{roots: make(map[string]*node)}
+}
+
+// newNode constructs a node.
+func newNode() *node {
+	return &node{
+		unenc:    make(map[string]*node),
+		unencMap: make(map[string]string),
+		enc:      make(map[string]*node),
+		encMap:   make(map[string]string),
+	}
+}
+
+// SetDefaultKey adds a default key to be returned for any lookup that does not match a bucket.
+// This is a temporary measure and should be removed when projects no longer require keys
+// to operate to encrypt buckets.
```

Comment:
is this still right?

Reply:
No, I guess not. I'll get this changed in a new PR to avoid blocking this.

Reply:
Other changes happened so this is happening now.